File: graph/shortest_bridge.java
public class shortest_bridge {
    
}
class Solution {
    public int shortestBridge(int[][] grid) {
        int N = grid.length;
        int[][] direct = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        
        Set<String> visit = new HashSet<>();
        
        // Depth First Search to mark one island cells
        Deque<int[]> queue = new ArrayDeque<>();
        
        // DFS function to mark one island cells
        // and put them into the visit set
        // We assume all cells are not visited yet.
        for (int r = 0; r < N; r++) {
            for (int c = 0; c < N; c++) {
                if (grid[r][c] == 1) {
                    dfs(grid, r, c, visit);
                    return bfs(grid, direct, visit);
                }
            }
        }
        return -1; // No solution found
    }
    
    // Depth First Search to mark one island cells
    private void dfs(int[][] grid, int r, int c, Set<String> visit) {
        int N = grid.length;
        if (r < 0 || c < 0 || r >= N || c >= N || grid[r][c] == 0 || visit.contains(r + "-" + c))
            return;
        
        visit.add(r + "-" + c);
        
        for (int[] dir : new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}) {
            dfs(grid, r + dir[0], c + dir[1], visit);
        }
    }
    
    // Breadth First Search to find the shortest bridge
    private int bfs(int[][] grid, int[][] direct, Set<String> visit) {
        int N = grid.length;
        Deque<int[]> queue = new ArrayDeque<>(visit.stream().map(s -> {
            String[] parts = s.split("-");
            return new int[]{Integer.parseInt(parts[0]), Integer.parseInt(parts[1])};
        }).collect(Collectors.toList()));
        
        int res = 0;
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] cell = queue.poll();
                int r = cell[0];
                int c = cell[1];
                for (int[] dir : direct) {
                    int newRow = r + dir[0];
                    int newCol = c + dir[1];
                    if (newRow < 0 || newRow >= N || newCol < 0 || newCol >= N || visit.contains(newRow + "-" + newCol))
                        continue;
                    
                    if (grid[newRow][newCol] == 1)
                        return res;
                    
                    queue.offer(new int[]{newRow, newCol});
                    visit.add(newRow + "-" + newCol);
                }
            }
            res++;
        }
        return -1; // No solution found
    }
}

File: graph/shortest_bridge.py
from typing import List
from collections import deque

class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        N = len(grid)
        direct = [[0,1],[0,-1],[1,0],[-1,0]]
        
        def invalid(r,c):
            return r < 0 or c < 0 or r == N or c == N
        visit = set()
        def dfs(r, c):
            if(invalid(r, c) or not grid[r][c] or (r, c) in visit):
                return 
            visit.add((r, c))
            for dr, dc in direct:
                dfs(r + dr,c + dc)
                
        def bfs():
            res, q = 0, deque(visit)
            while q:
                for _ in range(len(q)):
                    r, c = q.popleft()
                    for dr, dc in direct:
                        curR, curC = r + dr, c + dc
                        if invalid(curR, curC) or (curR, curC) in visit:
                            continue
                        if grid[curR][curC]:# reached 2nd island
                            return res
                        q.append([curR, curC])
                        visit.add((curR, curC))
                res += 1
                        
        for r in range(N):
            for c in range(N):
                if grid[r][c]:
                    dfs(r,c) # get one island cells in visit
                    # in bfs we check from which island cell we get shortest path to other island
                    return bfs()

'''
The `shortestBridge` function is designed to find the shortest path to connect two 
islands in a binary grid. Let’s analyze its time and space complexity:

### Time Complexity:
1. **DFS for Island Identification**:
   - The DFS function explores all cells of one island. In the worst case, this could 
   visit all \(N^2\) cells of the grid, where \(N\) is the number of rows or columns.
   - Thus, the time complexity for the DFS step is \(O(N^2)\).

2. **BFS for Shortest Path**:
   - After identifying one island, the BFS is initiated. The BFS could also potentially 
   visit all cells in the grid, and for each cell, it checks its 4 neighbors.
   - Therefore, the time complexity for the BFS step is also \(O(N^2)\).

3. **Overall Time Complexity**:
   - Since both the DFS and BFS could run independently through all \(N^2\) cells, 
   the overall time complexity is:
   \[
   O(N^2)
   \]

### Space Complexity:
1. **Visited Set**: 
   - The `visit` set stores cells of one island and possibly additional cells during 
   the BFS. In the worst case, it could contain up to \(N^2\) cells.
   - Therefore, the space complexity due to the visited set is \(O(N^2)\).

2. **Queue for BFS**:
   - The BFS uses a queue that can also hold up to \(O(N^2)\) cells at once in the 
   worst case.
   
3. **Overall Space Complexity**:
   - Thus, the overall space complexity is:
   \[
   O(N^2)
   \]

### Summary:
- **Time Complexity**: \(O(N^2)\)
- **Space Complexity**: \(O(N^2)\)
''' 